qiskit_experiments/analysis/curve_fitting.py

```diff
@@ -0,0 +1,291 @@
+            \sigma_i^{-2} (f(x_i, \Theta) -  y_i)^2
+
+    using ``scipy.optimize.curve_fit``.
+
+    Args:
+        func: a fit function `f(x, *params)`.
+        xdata: a 1D float array of x-data.
+        ydata: a 1D float array of y-data.
+        p0: initial guess for optimization parameters.
+        sigma: Optional, a 1D array of standard deviations in ydata.
+        bounds: Optional, lower and upper bounds for optimization
+                parameters.
+        kwargs: additional kwargs for scipy.optimize.curve_fit.
+
+    Returns:
+        result containing `popt` the optimal fit parameters,
+        `popt_err` the standard error estimates popt,
+        `pcov` the covariance matrix for the fit,
+        `chisq` the chi-squared parameter of fit,
+        `xrange` the range of xdata values used for fit.
+
+    Raises:
+        QiskitError: if the number of y-values in the dataset is not
+                     greater than the number of parameters in the
+                     fit function.
+    """
+    # Format p0 parameters if specified as dictionary
+    if isinstance(p0, dict):
+        param_keys = list(p0.keys())
+        param_p0 = list(p0.values())
+
+        # Convert bounds
+        if bounds:
+            lower = [bounds[key][0] for key in param_keys]
+            upper = [bounds[key][1] for key in param_keys]
+            param_bounds = (lower, upper)
+        else:
+            param_bounds = None
+
+        # Convert fit function
+        def fit_func(x, *params):
+            return func(x, **dict(zip(param_keys, params)))
+
+    else:
+        param_keys = None
+        param_p0 = p0
+        param_bounds = bounds
+        fit_func = func
+
+    # Check data set has at least 1 more values than model parameters
+    num_params = len(param_p0)
+    if len(ydata) <= num_params:
+        raise QiskitError(
+            "The number of y-values in the dataset is not greater than"
+            " the number of parameters in the fit function."
+        )
+
+    # Override scipy.curve_fit default for absolute_sigma to True
+    # if not specified in kwargs
+    if sigma is not None and "absolute_sigma" not in kwargs:
+        kwargs["absolute_sigma"] = True
+
+    # Run curve fit
+    # TODO: Add error handling so if fitting fails we can return an analysis
+    #       result containing this information
+    # pylint: disable = unbalanced-tuple-unpacking
+    popt, pcov = opt.curve_fit(
+        fit_func, xdata, ydata, sigma=sigma, p0=param_p0, bounds=param_bounds, **kwargs
+    )
+    popt_err = np.sqrt(np.diag(pcov))
+
+    # Calculate the reduced chi-squared for fit
+    yfits = fit_func(xdata, *popt)
+    residues = (yfits - ydata) ** 2
+    if sigma is not None:
+        residues = residues / (sigma ** 2)
+    chisq = np.sum(residues) / (len(yfits) - num_params)
+
+    # Compute xdata range for fit
+    xdata_range = [min(xdata), max(xdata)]
+
+    result = {
+        "popt": popt,
```

Comment:
question; if the user passed in a dict, should we return a dict here?

Reply:
To avoid having two different return types, i left this as an array, and if p0 was supplied with a dict the `popt_keys` will include the corresponding key labels for each param